### ISSUE TITLE
How to improve the speed of exe opening?

### ISSUE THREAD

**Issue by smp12345** [fyne-io/fyne#2015]
I have compiled a Fyne EXE in Win 10, which is about 20M in total, and it may take 2-3 seconds to open each time. Could you please speed up the startup?

**Comment by AlbinoGeek** [comment]
I am finding that a lot of my `fyne` applications take multiple seconds to open as well.

Would love to hear if plans will be made for tooling to allow developers to diagnose these startup times.

**Comment by Jacalz** [comment]
20 MB file sizes? That double what I’m getting. Are you sure that there isn’t a large amount of data packed into the library that needs to be loaded? I’m having very fast opening times on my end.

**Comment by AlbinoGeek** [comment]
Well hacking together a widget as root on the window and hooking the first call to `Refresh` or `Layout` as the "it finished loading" check, I get more than 6-10s on most program launch. If there were a supported way to know the startup and driver initialization times of the framework, that would be much more helpful for diagnosing this issue.

**Comment by Jacalz** [comment]
That’s very strange. Not that I run Fyne applications on Windows often, but on Linux, FreeBSD and MacOS load times seem to be not measurably slower than instant.

**Comment by andydotxyz** [comment]
Please be certain that you have asked your anti-virus to ignore the apps.
When I am running on windows It is snow with anti-virus running and very fast when it's not.
This seems to be a side-effect of compiling binaries locally, but I don't see why it does not impact apps from other languages.

**Comment by andydotxyz** [comment]
Just for information this is not a slowness within Fyne load per-se as most platforms are very fast.
macOS and Linux seem to be blazing fast (faster than GTK and Qt for sure).
Even loading apps on low powered devices such as PinePhone or RaspberryPi is still around 1/4 to 1/2 second

**Comment by AlbinoGeek** [comment]
Is there a test or benchmark that represents the startup ("time to first paint") and teardown time in the framework for the sake of things like this issue or regressions?

**Comment by andydotxyz** [comment]
> Is there a test or benchmark that represents the startup ("time to first paint") and teardown time

No there isn't, though it's a good idea, until now there had not been any mentions of it being slow so I guess we missed that.

**Comment by smp12345** [comment]
It is not practical to turn off the user's anti-virus software. Hope there is another good solution

**Comment by Jacalz** [comment]
I don't think we can have much control over what the antivirus does. My best guess would be to use `fyne release` and sign the binaries.

**Comment by andydotxyz** [comment]
The anti-virus should only impact load times for the first time a binary is run. Though they do vary.

**Comment by aksdb** [comment]
If I open the fyne_demo on my linux machine it also takes ~1s to show the window. If I open - in comparison - a gioui application (so it's also Go and custom rendered UI using OpenGL) it shows up instantly. So I don't think this is Windows specific.

**Comment by egonelbre** [comment]
Recent Go version support `GODEBUG=inittrace=1`  https://github.com/golang/go/commit/7c58ef732efd9bf0d0882bb95371ce1909924a75. This should show whether some init func is taking time to load or is the time spent somewhere else.

**Comment by egonelbre** [comment]
I just remembered that one issue I have seen is with `github.com/spf13/cobra`, which has a mousetrap that causes a slow startup. https://github.com/spf13/cobra/blob/master/command_win.go

**Comment by egonelbre** [comment]
The only slow initialization I found in fyne_demo was github.com/fredbi/uri dependency due to these regular expressions at https://github.com/fredbi/uri/blob/3dcfdacbaaf3aeab42eefc96820857792fb55a1f/uri.go#L330.

**Comment by Bene-GG** [comment]
After some testing I found out that apps made using fyne only need like ~8 seconds to start while the TeamSpeak 5 Beta client is running. Here's a list of the used components inside the TS5 client.
![listofts5deps](https://user-images.githubusercontent.com/8401134/119193359-ca267e80-ba81-11eb-8e29-3c480ba06d21.png)

**Comment by egonelbre** [comment]
@Bene-GG those components are unlikely to cause these problems. I would guess that the issue is that TS5 does some sort of hooking into the program -- shortcut keys, OpenGL or something like that -- and that's what's taking the time. And it's quite likely that there are other programs that do it as well.

**Comment by andydotxyz** [comment]
> The other part that took long was
> 
> https://github.com/fyne-io/fyne/blob/a25b08480d742f016ee24c1b37aa02c3bb433bd5/internal/driver/glfw/loop.go#L70
> 
> (with around 300ms)

Yes this is the only part that is unavoidable (with the way driver is currently implemented).

I don't think we can do much about TS5 intercepting our app and/or antivirus slowing launch. Should this ticket be closed as invalid?

**Comment by Jacalz** [comment]
The uri package was improved to init faster and we have had new GLFW versions since this was opened so likely not an issue any more.
